lib/Db/MessageMapper.php

```diff
@@ -954,28 +945,33 @@ public function findRelatedData(array $messages, string $userId): array {
 	 * @return int[]
 	 */
 	public function findNewIds(Mailbox $mailbox, array $ids): array {
-		$sub = $this->db->getQueryBuilder();
-		$qb = $this->db->getQueryBuilder();
+		$select = $this->db->getQueryBuilder();
+		$subSelect = $this->db->getQueryBuilder();
+
+		$knowIdsParameter = $select->createParameter('ids');
 
-		$subSelect = $sub
-			->select($sub->func()->max('uid'))
+		$subSelect
+			->select($subSelect->func()->min('sent_at'))
 			->from($this->getTableName())
 			->where(
-				$sub->expr()->eq('mailbox_id', $qb->createNamedParameter($mailbox->getId(), IQueryBuilder::PARAM_INT)),
-				$sub->expr()->in('id', $qb->createParameter('ids'))
+				$subSelect->expr()->eq('mailbox_id', $select->createNamedParameter($mailbox->getId(), IQueryBuilder::PARAM_INT)),
+				$subSelect->expr()->in('id', $knowIdsParameter)
 			);
 
-		$qb->select('id')
-			->from($this->getTableName())
+		$select
+			->select('m.id')
+			->from($this->getTableName(), 'm')
+			->leftJoin('m', $this->getTableName(), 'm2', 'm.mailbox_id = m2.mailbox_id and m.thread_root_id = m2.thread_root_id and m.sent_at < m2.sent_at')
 			->where(
-				$qb->expr()->eq('mailbox_id', $qb->createNamedParameter($mailbox->getId(), IQueryBuilder::PARAM_INT))
-			);
+				$select->expr()->eq('m.mailbox_id', $select->createNamedParameter($mailbox->getId(), IQueryBuilder::PARAM_INT)),
+				$select->expr()->notIn('m.id', $knowIdsParameter),
+				$select->expr()->gt('m.sent_at', $select->createFunction('(' . $subSelect->getSQL() . ')'), IQueryBuilder::PARAM_INT),
+				$select->expr()->isNull('m2.id')
+			)
+			->orderBy('m.sent_at', 'desc');
 
-		return array_flat_map(function (array $chunk) use ($qb, $subSelect) {
-			$qb->setParameter('ids', $chunk, IQueryBuilder::PARAM_INT_ARRAY);
-			$select = $qb->andWhere(
-				$qb->expr()->gt('uid', $qb->createFunction('(' . $subSelect->getSQL() . ')'), IQueryBuilder::PARAM_INT)
-			);
+		return array_flat_map(function (array $chunk) use($select) {
```

Comment:
That does not work. The second chunk will return all id's from the first chunk as new. I have some ideas how to workaround:

1) `array_diff(array_flat_map(....), $ids);`.

2) Instead of using one in/notIn expression split the ids into 1000er chunks and conjunction/disjunction the in/notIn statments.

Code:
```
		$select = $this->db->getQueryBuilder();
		$subSelect = $this->db->getQueryBuilder();
		
		$inExpression = [];
		$notInExpression = [];

		foreach (array_chunk($ids, 1000) as $chunk) {
			$inExpression[] = $subSelect->expr()->in('id', $select->createNamedParameter($chunk, IQueryBuilder::PARAM_INT_ARRAY));
			$notInExpression[] = $subSelect->expr()->notIn('m.id', $select->createNamedParameter($chunk, IQueryBuilder::PARAM_INT_ARRAY));
		}

		$subSelect
			->select($subSelect->func()->min('sent_at'))
			->from($this->getTableName())
			->where(
				$subSelect->expr()->eq('mailbox_id', $select->createNamedParameter($mailbox->getId(), IQueryBuilder::PARAM_INT)),
				$subSelect->expr()->orX(...$inExpression)
			);

		$select
			->select('m.id')
			->from($this->getTableName(), 'm')
			->leftJoin('m', $this->getTableName(), 'm2', 'm.mailbox_id = m2.mailbox_id and m.thread_root_id = m2.thread_root_id and m.sent_at < m2.sent_at')
			->where(
				$select->expr()->eq('m.mailbox_id', $select->createNamedParameter($mailbox->getId(), IQueryBuilder::PARAM_INT)),
				$select->expr()->andX(...$notInExpression),
				$select->expr()->gt('m.sent_at', $select->createFunction('(' . $subSelect->getSQL() . ')'), IQueryBuilder::PARAM_INT),
				$select->expr()->isNull('m2.id')
			)
			->orderBy('m.sent_at', 'desc');

		return $this->findIds($select);
```

Query:
```
SELECT `m`.`id`
FROM `*PREFIX*mail_messages` `m`
         LEFT JOIN `*PREFIX*mail_messages` `m2`
                   ON m.mailbox_id = m2.mailbox_id and
                      m.thread_root_id = m2.thread_root_id and
                      m.sent_at < m2.sent_at
WHERE (`m`.`mailbox_id` = :dcValue12)
  AND ((`m`.`id` NOT IN (:dcValue2)) AND (`m`.`id` NOT IN (:dcValue4)) AND
       (`m`.`id` NOT IN (:dcValue6)) AND (`m`.`id` NOT IN (:dcValue8)) AND
       (`m`.`id` NOT IN (:dcValue10)))
  AND (`m`.`sent_at` > (SELECT MIN(`sent_at`)
                        FROM `*PREFIX*mail_messages`
                        WHERE (`mailbox_id` = :dcValue11)
                          AND ((`id` IN (:dcValue1)) OR (`id` IN (:dcValue3)) OR
                               (`id` IN (:dcValue5)) OR (`id` IN (:dcValue7)) OR
                               (`id` IN (:dcValue9)))))
  AND (`m2`.`id` IS NULL)
ORDER BY `m`.`sent_at` desc
```